common/src/abi/mod.rs

```diff
@@ -47,12 +50,54 @@ pub struct Abi {
 pub type MethodNum = u64;
 
 /// `Methods` describes an exposed method from an actor entrypoint.
-#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Serialize_tuple, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(crate = "fvm_ipld_encoding::serde")]
 pub struct Method {
+    #[serde(skip_serializing)]
     pub number: MethodNum,
     pub name: String,
 }
 
+impl<'de> fvm_ipld_encoding::de::Deserialize<'de> for Method {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: fvm_ipld_encoding::de::Deserializer<'de>,
+    {
+        struct MethodVisitor;
+
+        impl<'de> fvm_ipld_encoding::de::Visitor<'de> for MethodVisitor {
+            type Value = Method;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("Method")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> std::result::Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let name = seq
+                    .next_element::<String>()?
+                    .ok_or_else(|| fvm_ipld_encoding::de::Error::invalid_length(0, &self))?;
+
+                let number = match derive_method_num(&name) {
+                    Ok(number) => number,
+                    Err(_) => {
+                        return Err(fvm_ipld_encoding::de::Error::custom(format_args!(
+                            "Could not derive method number for: {}",
+                            &name
+                        )))
+                    }
+                };
+
+                Ok(Self::Value { number, name })
+            }
+        }
+
+        deserializer.deserialize_seq(MethodVisitor)
+    }
+}
```

Comment:
wdyt of using https://github.com/Keats/validator instead, and have:
```Rust
pub struct Method {
    #[validate(custom = "validate_method_num")]
    #[serde(skip_serializing)]
    pub number: MethodNum,
    pub name: String,
}

fn validate_method_num(num: MethodNum) -> Result<(), ValidationError> {
    derive_method_num(num).map_err(|err| ValidationError::new(&err.to_string()))
}
```

we'd then not require all this code and its tests.

Reply:
- The `validator` package did not work as intended as there is a need to call `validate` to check fields while we want to directly end up with a structure that contains both a `name` and a `number`
- Also tried to use `#[serde(deserialize_with = "path")]` but i was also a dead end. The way that deserialization of CBOR works is that it first checks for the length of the next tuple of values in the bytes before deserializing it in our Rust structure. It is at this step that it throws an error as it expects a  tuple of 2 values (`name` and `number`) but receives only one (`name`).

So I know that it would be great to simplify but I think that the current code is valid and lets us do what we want.

Reply:
Yeah makes sense Thomas, thanks for investigating it. I'd just recommend leaving a TODO: with the improvement note :)